### ISSUE TITLE
Add key `failOnStderr` to yaml pipeline schema

### ISSUE THREAD

**Issue by svdHero** [woodpecker-ci/woodpecker#1465]
### Clear and concise description of the problem

As a user of Woodpecker I would like to be able to set a boolean parameter `failOnStderr` for each pipeline step that makes the step fail if any of the listed commands print anything to stderr. In some scenarios that can be useful for catching warnings that are not reflected by a command's exit code.

Right now, the following pipeline succeeds:

```yml
pipeline:
  my-step:
    image: debian:bullseye-slim
    commands:
      - echo "Do one thing."
      - echo "Booooom. Bad error!" 1>&2
      - echo "Do something else."
```

### Suggested solution

Add the boolean parameter `failOnStderr` to the yaml pipeline schema, such that
```yml
pipeline:
  my-step:
    image: debian:bullseye-slim
    failOnStderr: false
    commands:
      - echo "Do one thing."
      - echo "Booooom. Bad error!" 1>&2
      - echo "Do something else."
```
would succeed, but
```yml
pipeline:
  my-step:
    image: debian:bullseye-slim
    failOnStderr: true
    commands:
      - echo "Do one thing."
      - echo "Booooom. Bad error!" 1>&2
      - echo "Do something else."
```
would fail.

Also compare the [Bash Task](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/bash-v3?view=azure-pipelines) for Azure DevOps Pipelines which offers this feature.

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by gapodo** [comment]
IMHO this should be handled by the binary / script being executed, as it's non-standard behavior to exit with 0 when an error occurred. Other ways just wrap the binary in a script that catches stderr and if not empty exit with 1 (or what ever you'd like the code to be).

Is there any special use case you have in mind, where this would help with a standard compliant binary / script?

**Comment by smainz** [comment]
Comparing to what you cited regarding the Bash Task this can be handled by a plugin which does the wrapping @gapodo mentioned.

**Comment by svdHero** [comment]
> IMHO this should be handled by the binary / script being executed, as it's non-standard behavior to exit with 0 when an error occurred. Other ways just wrap the binary in a script that catches stderr and if not empty exit with 1 (or what ever you'd like the code to be).
> 
> Is there any special use case you have in mind, where this would help with a standard compliant binary / script?

There are quite a few programs out there that write their warnings to stderr but return an exit code of 0, because 1 is reserved for critical errors. However, in most use cases the pipeline should also fail in case of warnings.

One such program would be the PostgreSQL client psql. When there are certain errors in processing your SQL script, psql would return exit code 0, but write warnings to stderr. I have had that multiple times now.

I don't like hiding my pipeline code in wrapper shell scripts unnecessarily. I find it much more transparent to see in the central `.woodpecker.yml` which tool chain commands are triggered. But that's a matter of taste, I guess.

**Comment by svdHero** [comment]
> Comparing to what you cited regarding the Bash Task this can be handled by a plugin which does the wrapping @gapodo mentioned.

But that would mean that I have to write a wrapper plugin for each and every base image that I want to use as a build step, right?

**Comment by gapodo** [comment]
> But that would mean that I have to write a wrapper plugin for each and every base image that I want to use as a build step, right?

Theoretically yes, but after just a tiny bit of Google foo I found the `ON_ERROR_STOP` flag for your `psql example, while having to activate something to fail if there is an error is non-compliant at least it's easy to set... so using `psql -v ON_ERROR_STOP=on <other stuff goes here>` or adding `\set ON_ERROR_STOP on` to `~/.psqlrc` would perform as expected (exits with 3 on the first error). Keep in mind though that failing SQL can cause a bit of a mess, so --single-transaction is often needed to properly fail on non wiped services.



> I don't like hiding my pipeline code in wrapper shell scripts unnecessarily. I find it much more transparent to see in the central `.woodpecker.yml` which tool chain commands are triggered. But that's a matter of taste, I guess.

I fully understand that. Since the Linux shells feel a bit like LEGO for grown-ups, it's nice not to hide your bricks, the problem is when you have bricks that don't match the standard, and you have to integrate them (think stuff like batteries in LEGO) you'll need to somehow wrap it to make it compliant, either setting an option if that's possible (like with psql) of by wrapping it (as may sometimes be needed with badly done 3rd party stuff).

---

If you have any more examples of applications not properly failing with an exit code `!= 0` I'd be glad to know (it's just that I haven't stumbled across many, except some of my old hacky scripts), since knowing the prevalence of binaries / scripts not following the standard would allow us to better estimate the real world impact.

**Comment by gapodo** [comment]
Another problem that introducing this flag would have from an implementation standpoint is that once users / pipelines may choose the shell being used (#278, it's somewhat possible now by changing the entrypoint of an image), this would have to be built up fairly generically and tested with a large amount of shells (just some that come to mind quickly: sh, bash, dash, csh, fish, zsh, busybox,... I'm sure there are way more commonly in use, and some have multiple versions which are not fully compatible), or outsourced to the log-collection (which is extremely suboptimal since it should not be the concern of an async collector, but directly in the execution thereby ensuring immediate cancelation).

Also we have to keep in mind that there is not only the docker backend, but also Kubernetes (which is fairly similar to docker since they both use OCI images), ssh and local (maybe I'm missing some?)

**Comment by svdHero** [comment]
@gapodo thanks for the googling. I am sure I tried that flag, but still experienced some cases where the pioeline did not fail. However, I cannot pinpoint them right now.

Another candidate would be Azure CLI, but I know that there have been Microsoft efforts to align exit code with stderr output.

**Comment by 6543** [comment]
we already have an config to ignore errors ...
... handling specific non posix compliant suff is outside of the scope I would say ...

keep in mind -> KISS

I'm closing this issue, as it wont be added because of unnecessary complexity for a to specific issue, witch can be handled with bit shell magic.

tldr: if the tool does not work right, that one should be fixed instead

**Comment by 6543** [comment]
still thanks for the input, it's always appreciated :+1:
